src/vcpkg/commands.install.cpp

```diff
@@ -40,130 +44,202 @@ namespace vcpkg
         return dirs;
     }
 
-    const Path& InstallDir::destination() const { return this->m_destination; }
+    struct PathAndType
+    {
+        Path path;
+        FileType type = FileType::none;
+
+        bool operator<(const PathAndType& other) const noexcept
+        {
+            if (!path.empty() && other.path.empty())
+            {
+                return true;
+            }
+            if (path.empty() && !other.path.empty())
+            {
+                return false;
+            }
+            return path < other.path;
+        }
+        bool operator==(const PathAndType& other) const noexcept { return path == other.path && type == other.type; }
+    };
+
+    static std::vector<PathAndType> filter_files_to_install(const Filesystem& fs, std::vector<Path>&& files)
+    {
+        std::vector<PathAndType> output(files.size());
+        static const PathAndType invalid{"", FileType::none};
+
+        parallel_transform(files, output.begin(), [&](auto&& file) -> PathAndType {
+            std::error_code ec;
+            auto status = fs.symlink_status(file, ec);
```

Comment:
I strongly suspect the only thing you're gaining from parallelism here is parallelizing this `symlink_status` and this would be much clearer if you only parallelized that rather than trying to make invalid states that need to be chopped off etc.

If you parallelize just this part you could use plain `Util::erase_remove_if` for the invalid ones. It would also have the advantage of not trying to parallelize `println_warning`.

Reply:
If I want to get rid of invalid states I'd have to use `parallel_for_each` but this would require a lock on when inserting things into the destination vector. The only thing I can move out of this is the validation of the file type.